packages/jupyterlab-manager/src/semvercache.ts

```diff
@@ -3,22 +3,54 @@
 
 import { maxSatisfying } from 'semver';
 
+import { PromiseDelegate } from '@lumino/coreutils';
+
+import { ArrayExt } from '@lumino/algorithm';
+
+const TIMEOUT = 500;
+
+type TPromiseCache<T> = [
+  string,
+  string,
+  PromiseDelegate<T | undefined>,
+  number | null
+];
+
 /**
  * A cache using semver ranges to retrieve values.
  */
 export class SemVerCache<T> {
-  set(key: string, version: string, object: T): void {
-    if (!(key in this._cache)) {
-      this._cache[key] = Object.create(null);
+  _promiseDelegates: TPromiseCache<T>[] = [];
+  _timeout: number;
+
+  set(options: SemVerCache.IRegistryData<T>): void {
+    const { name, version, object } = options;
+    if (!(name in this._cache)) {
+      this._cache[name] = Object.create(null);
     }
-    if (!(version in this._cache[key])) {
-      this._cache[key][version] = object;
+    if (!(version in this._cache[name])) {
+      this._cache[name][version] = object;
     } else {
-      throw `Version ${version} of key ${key} already registered.`;
+      throw `Version ${version} of name ${name} already registered.`;
     }
+    this._retry();
   }
 
-  get(key: string, semver: string): T | undefined {
+  _retry(): void {
+    clearTimeout(this._timeout);
```

Comment:
Can the pollers in lumino help with this logic?

Reply:
Thanks for checking out this PR! :) Sounds like they could!

Reply:
hmm .. i don't know... our logic is pretty simple and the functions are quite straight forward. I don't know if using the polling api would make it more easy to understand -- and also I do not understand which part I should really use.

Reply:
We just thought this through, and we think a small change in the logic would let us use the throttler without too much confusion or hassle.

Reply:
Done!